### PR TITLE
Bookindex

### PR DESCRIPTION
**Pull request by aleksandraziemak**
BookRepository improved

### CODE REVIEW

File: src/main/java/com/library/BookRepository.java [hunk 1]
```diff
@@ -59,29 +64,36 @@ private void writeToJsonFile(JSONArray jsonBooks) {
...
     }
 
-    public void settingNewAuthor(int index, Author author) {
+    public void settingNewAuthor(Book book, Author author) {
         List<Book> books = getBooksFromJsonFile();
-        books.get(index).setAuthor(author);
+        books.remove(book);
+        book.setAuthor(author);
+        books.add(book);
         JSONArray jsonBooks = jsonBooks(books);
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java]: jsonBooks -> stwórz nową klasę BookMapper z funkcją statyczną "mapBooks"

File: src/main/java/com/library/BookRepository.java [hunk 2]
```diff
@@ -94,7 +106,7 @@ public List<Book> getBooksFromJsonFile() {
                 JSONObject book = (JSONObject) object; //wskazuję, że obiekt book jest obiektem Json'owym
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java]: te 4 linijki też wynieś do mappera

File: src/main/java/com/library/BookService.java [hunk 3]
```diff
@@ -108,43 +122,49 @@ public void deleteBook() {
...
     public void editBook() {
         System.out.println("Edit a book in Library");
-        showMyLibrary();
+        List<Book> books = repository.getBooksFromJsonFile();
+        showMyLibrary(books);
         System.out.println("Please, choose number (book to edit):");
         int index = getIndex();
-        System.out.println("Do you want to change Title of a book: " + repository.getBook(index).getTitle() + "? (y / n)");
+        Book book = books.get(index);
+        System.out.println("Do you want to change Title of a book: " + repository.getBook(book).getTitle() + "? (y / n)");
```
Comment (jandrusz) [inline src/main/java/com/library/BookService.java]: repository.getBook(book).getTitle() -> book.getTitle()

File: src/main/java/com/library/BookService.java [hunk 4]
```diff
@@ -108,43 +122,49 @@ public void deleteBook() {
...
         String option = getOption();
         editingTitle(option, index);
-        System.out.println("Do you want to change Author of a book: " + repository.getBook(index).getAuthor().getFullName() + "? (y / n)");
+        System.out.println("Do you want to change Author of a book: " + repository.getBook(book).getAuthor().getFullName() + "? (y / n)");
```
Comment (jandrusz) [inline src/main/java/com/library/BookService.java]: repository.getBook(book).getTitle() -> book.getAuthor()

File: src/main/java/com/library/BookRepository.java [hunk 5]
```diff
@@ -59,29 +64,36 @@ private void writeToJsonFile(JSONArray jsonBooks) {
...
+        books.remove(book);
+        book.setAuthor(author);
+        books.add(book);
         JSONArray jsonBooks = jsonBooks(books);
         writeToJsonFile(jsonBooks);
     }
 
-    public Book getBook(int index) {
-        return getBooksFromJsonFile().get(index);
+    public Book getBook(Book book) {
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java]: chyba niepotrzebne jeżeli poprawisz dwa pozostałe komentarze?

File: src/main/java/com/library/BookRepository.java (hunk 1 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookRepository.java]: Done

File: src/main/java/com/library/BookRepository.java (hunk 2 above)
Comment (jandrusz) [reply src/main/java/com/library/BookRepository.java]: wynikowo ten kawałek chce taki:
for (Object object : booksJson) {
books.add(BookMapper.map((JSONObject) object));
}

File: src/main/java/com/library/BookMapper.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import com.library.model.Book;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+
+import java.util.List;
+import java.util.UUID;
+
+public class BookMapper {
+
+    private static final BookRepository repository = new BookRepository();
```
Comment (jandrusz) [inline src/main/java/com/library/BookMapper.java]: do wywalenia

File: src/main/java/com/library/BookMapper.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        return jsonBooks;
+/*        for (Book book: books) {
+            JSONObject newBook = new JSONObject();
+            newBook.put("title", book.getTitle());
+            newBook.put("author", book.getAuthor());
+            jsonBooks.add(newBook);
+        }*/
+    }
+
+    public static Book mappedBook(JSONObject object) {
```
Comment (jandrusz) [inline src/main/java/com/library/BookMapper.java]: mappedBook -> mapJson

File: src/main/java/com/library/BookRepository.java [hunk 8]
```diff
@@ -19,36 +22,16 @@ public class BookRepository {
     private static final String LAST_NAME = "lastName";
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java]: do wywalenia:

    private static final String FIRST_NAME = "firstName";
    private static final String LAST_NAME = "lastName";
    private static final String TITLE = "title";
    private static final String AUTHOR = "author";
    private static final String ID = "id";

File: src/main/java/com/library/BookMapper.java (hunk 7 above)
Comment (jandrusz) [reply src/main/java/com/library/BookMapper.java]: public static Book mappedBook(JSONObject book) {
        JSONObject jsonAuthor = (JSONObject) book.get(AUTHOR);
        Author author = new Author((String) jsonAuthor.get(FIRST_NAME), (String) jsonAuthor.get(LAST_NAME));
        return new Book((String) book.get(TITLE), author, UUID.fromString((String) book.get(ID)));
    }

File: src/main/java/com/library/BookRepository.java [hunk 9]
```diff
@@ -59,42 +42,40 @@ private void writeToJsonFile(JSONArray jsonBooks) {
...
-            FileReader fr = new FileReader(PATH_TO_JSON_FILE); //wskazuję plik
-            JSONArray booksJson = (JSONArray) parser.parse(fr); //wskazuję plik jako Json'owy plik i robię z niego Array
-            for (Object object : booksJson) { //idę po kolei po wszystkich obiektach z listy booksJson
-                JSONObject book = (JSONObject) object; //wskazuję, że obiekt book jest obiektem Json'owym
-                JSONObject jsonAuthor = (JSONObject) book.get(AUTHOR);
-                Author author = new Author((String) jsonAuthor.get(FIRST_NAME), (String) jsonAuthor.get(LAST_NAME));
-                books.add(new Book((String) book.get(TITLE), author)); //dodaję new Book do listy books, na podstawie obiektów Json'owych
+            FileReader fr = new FileReader(PATH_TO_JSON_FILE);
+            JSONArray booksJson = (JSONArray) parser.parse(fr);
+            for (Object object : booksJson) {
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java]: booksJson.forEach(object -> books.add(BookMapper.mappedBook((JSONObject) object)));

File: src/main/java/com/library/BookRepository.java [hunk 10]
```diff
@@ -19,36 +22,16 @@ public class BookRepository {
...
-                })
-                .forEach(jsonObj -> jsonBooks.add(jsonObj));
-        return jsonBooks;
-/*        for (Book book: books) {
-            JSONObject newBook = new JSONObject();
-            newBook.put("title", book.getTitle());
-            newBook.put("author", book.getAuthor());
-            jsonBooks.add(newBook);
-        }*/
-    }
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java]: usuń pustą linię

File: src/main/java/com/library/BookRepository.java [hunk 11]
```diff
@@ -59,42 +42,40 @@ private void writeToJsonFile(JSONArray jsonBooks) {
...
+    public void settingNewAuthor(Book book, Author author) {
         List<Book> books = getBooksFromJsonFile();
-        books.get(index).setAuthor(author);
-        JSONArray jsonBooks = jsonBooks(books);
+        books.remove(book);
+        book.setAuthor(author);
+        books.add(book);
+        JSONArray jsonBooks = BookMapper.mapBooks(books);
         writeToJsonFile(jsonBooks);
     }
```
Comment (jandrusz) [inline src/main/java/com/library/BookRepository.java:62]: usuń pustą linię

File: src/main/java/com/library/BookMapper.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    private static final BookRepository repository = new BookRepository();
+    private static final String FIRST_NAME = "firstName";
+    private static final String LAST_NAME = "lastName";
+    private static final String TITLE = "title";
+    private static final String AUTHOR = "author";
+    private static final String ID = "id";
+
+    public static JSONArray mapBooks(List<Book> books) {
+        JSONArray jsonBooks = new JSONArray();
+        books.stream()
```
Comment (jandrusz) [inline src/main/java/com/library/BookMapper.java:22]: public static JSONArray mapBooks(List<Book> books) {
        JSONArray jsonBooks = new JSONArray();
        books.stream()
            .map(BookMapper::mapBook)
            .forEach(jsonBooks::add);
        return jsonBooks;
/*        for (Book book: books) {
            JSONObject newBook = new JSONObject();
            newBook.put("title", book.getTitle());
            newBook.put("author", book.getAuthor());
            jsonBooks.add(newBook);
        }*/
    }

    private static JSONObject mapBook(Book book) {
        JSONObject newBook = new JSONObject();
        newBook.put(TITLE, book.getTitle());
        newBook.put(ID, book.getId().toString());
        JSONObject newAuthor = new JSONObject();
        newAuthor.put(FIRST_NAME, book.getAuthor().getFirstName());
        newAuthor.put(LAST_NAME, book.getAuthor().getLastName());
        newBook.put(AUTHOR, newAuthor);
        return newBook;
    }

File: src/main/java/com/library/BookMapper.java (hunk 6 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookMapper.java]: +

File: src/main/java/com/library/BookMapper.java (hunk 12 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookMapper.java:22]: +

File: src/main/java/com/library/BookMapper.java (hunk 7 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookMapper.java]: +

File: src/main/java/com/library/BookRepository.java (hunk 8 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookRepository.java]: +

File: src/main/java/com/library/BookRepository.java (hunk 10 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookRepository.java]: +

File: src/main/java/com/library/BookRepository.java (hunk 11 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookRepository.java:62]: +

File: src/main/java/com/library/BookRepository.java (hunk 9 above)
Comment (aleksandraziemak) [reply src/main/java/com/library/BookRepository.java]: +

File: src/main/java/com/library/BookMapper.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        newBook.put(TITLE, book.getTitle());
+        newBook.put(ID, book.getId().toString());
+        JSONObject newAuthor = new JSONObject();
+        newAuthor.put(FIRST_NAME, book.getAuthor().getFirstName());
+        newAuthor.put(LAST_NAME, book.getAuthor().getLastName());
+        newBook.put(AUTHOR, newAuthor);
+        return newBook;
+    }
+
+    public static Book mapJson(JSONObject object) {
```
Comment (jandrusz) [inline src/main/java/com/library/BookMapper.java]: public static Book mapJson(JSONObject book) {
        JSONObject jsonAuthor = (JSONObject) book.get(AUTHOR);
        Author author = new Author((String) jsonAuthor.get(FIRST_NAME), (String) jsonAuthor.get(LAST_NAME));
        return new Book((String) book.get(TITLE), author, UUID.fromString((String) book.get(ID)));
    }

**Pull request closed without merging**